routing/cross_mwm_road_graph.cpp

```diff
@@ -129,24 +161,35 @@ IRouter::ResultCode CrossMwmGraph::SetFinalNode(CrossNode const & finalNode)
   return IRouter::NoError;
 }
 
-vector<BorderCross> const & CrossMwmGraph::ConstructBorderCross(OutgoingCrossNode const & startNode,
-                                                                TRoutingMappingPtr const & currentMapping) const
+vector<BorderCross> const & CrossMwmGraph::ConstructBorderCrossByOutgoing(OutgoingCrossNode const & startNode,
+                                                                          TRoutingMappingPtr const & currentMapping) const
 {
-  // Check cached crosses.
   auto const key = make_pair(startNode.m_nodeId, currentMapping->GetMwmId());
-  auto const it = m_cachedNextNodes.find(key);
-  if (it != m_cachedNextNodes.end())
+  auto const it = m_cachedNextNodesByOutgoing.find(key);
+  if (it != m_cachedNextNodesByOutgoing.end())
     return it->second;
 
-  // Cache miss case.
-  auto & crosses = m_cachedNextNodes[key];
-  ConstructBorderCrossImpl(startNode, currentMapping, crosses);
+  auto & crosses = m_cachedNextNodesByOutgoing[key];
+  ConstructBorderCrossByOutgoingImpl(startNode, currentMapping, crosses);
   return crosses;
 }
 
-bool CrossMwmGraph::ConstructBorderCrossImpl(OutgoingCrossNode const & startNode,
-                                             TRoutingMappingPtr const & currentMapping,
-                                             vector<BorderCross> & crosses) const
+vector<BorderCross> const & CrossMwmGraph::ConstructBorderCrossByIngoing(IngoingCrossNode const & startNode,
+                                                                         TRoutingMappingPtr const & currentMapping) const
+{
+  auto const key = make_pair(startNode.m_nodeId, currentMapping->GetMwmId());
+  auto const it = m_cachedNextNodesByIngoing.find(key);
+  if (it != m_cachedNextNodesByIngoing.end())
+    return it->second;
+
+  auto & crosses = m_cachedNextNodesByIngoing[key];
+  ConstructBorderCrossByIngoingImpl(startNode, currentMapping, crosses);
```

Comment:
Note, that you may pass caches and pointers to methods to the third common method.